internal/pkg/service/stream/definition/switchable.go

```diff
@@ -7,48 +7,109 @@ import (
 )
 
 type SwitchableInterface interface {
-	Enable()
-	Disable(now time.Time, by, reason string)
+	// Enable marks the entity as enabled.
+	Enable(now time.Time, by By)
+	// Disable marks the entity as disabled.
+	Disable(now time.Time, by By, reason string, disabledWithParent bool)
+	// IsEnabled returns true if the entity is marked as enabled.
 	IsEnabled() bool
-	GetDisabledBy() string
-	GetDisabledAt() *utctime.UTCTime
-	GetDisabledReason() string
+	IsEnabledAt(at time.Time) bool
+	// IsDisabled returns true if the entity is marked as disabled.
+	IsDisabled() bool
+	IsDisabledAt(at time.Time) bool
+	// IsDisabledDirectly returns true if the entity has been disabled directly, not together with its parent.
+	IsDisabledDirectly() bool
+	DisabledBy() *By
+	DisabledAt() *utctime.UTCTime
+	DisabledReason() string
+	EnabledBy() *By
+	EnabledAt() *utctime.UTCTime
 }
 
 type Switchable struct {
-	Disabled       bool             `json:"disabled,omitempty"`
-	DisabledBy     string           `json:"disabledBy,omitempty" validate:"required_if=Disabled true,excluded_if=Disabled false"`
-	DisabledAt     *utctime.UTCTime `json:"disabledAt,omitempty"  validate:"required_if=Disabled true,excluded_if=Disabled false"`
-	DisabledReason string           `json:"disabledReason,omitempty"  validate:"required_if=Disabled true,excluded_if=Disabled false"`
+	Disabled *Disabled `json:"disabled,omitempty" validate:"excluded_with=Enabled"`
+	Enabled  *Enabled  `json:"enabled,omitempty" validate:"excluded_with=Disabled"`
+}
+
+type Disabled struct {
+	// Directly is true if the entity has been disabled directly, not together with its parent.
+	Directly bool            `json:"directly"`
+	At       utctime.UTCTime `json:"at" validate:"required"`
+	Reason   string          `json:"reason" validate:"required"`
+	By       By              `json:"by" validate:"required"`
+}
+
+type Enabled struct {
+	At utctime.UTCTime `json:"at" validate:"required"`
+	By By              `json:"by" validate:"required"`
+}
+
+func (v *Switchable) Enable(now time.Time, by By) {
+	v.Disabled = nil
+	v.Enabled = &Enabled{At: utctime.From(now), By: by}
+}
+
+func (v *Switchable) Disable(now time.Time, by By, reason string, directly bool) {
+	v.Disabled = &Disabled{At: utctime.From(now), By: by, Reason: reason, Directly: directly}
+	v.Enabled = nil
 }
 
 func (v *Switchable) IsEnabled() bool {
-	return !v.Disabled
+	return v.Disabled == nil
```

Comment:
I would couple `SomethingEnabled` with  interaction with `v.Enabled` and `SomethingDisabled` functions with `v.Disabled`. This brakes it.

Reply:
added comment https://github.com/keboola/keboola-as-code/pull/1774/commits/1b1248ea0d813c8827a1777b3517e061e0ee4891